shell/components/CruResource.vue

```diff
@@ -324,7 +325,10 @@ export default {
 
     async showPreviewYaml() {
       if ( this.applyHooks ) {
-        await this.applyHooks(BEFORE_SAVE_HOOKS);
+        await this.applyHooks(
+          BEFORE_SAVE_HOOKS,
+          { _omitHooksMarker: true, names: ['save-machine-pools'] }
```

Comment:
This brings in a resource specific reference to a generic component. In the future i think the we'll split out provisioning into extensions and this component will move to rancher/components.

Can the applyHooks fn accept a context (for this is would be something like `show-preview-yaml`) that is passed to the hook when called. The individual hook can then choose to exit early given that context.

Reply:
> This brings in a resource specific reference to a generic component

Agreed, that's not desirable. I added a new `omitHooks` prop to address the problem, but that still doesn't feel like the correct abstraction. The new prop doesn't do a good job of explaining what are we omitting hooks for.

> Can the applyHooks fn accept a context (for this is would be something like show-preview-yaml) that is passed to the hook when called. The individual hook can then choose to exit early given that context.

We can, but we would still have the problem of needing to specify which hooks need to be omitted. Otherwise, we'll never know when to exit early.. 

Unless we think that hooks should never be applied when clicking the "Edit as YAML" button. If that's the case, we could just remove `applyHooks` from the button altogether.

Reply:
With the suggested change, all hooks would still run but each hook would decide if it needed to do anything given the context supplied to it. Which i think is basically below

```
const showPreviewYamlContext = 'show-preview-yaml`

async showPreviewYaml() {
  if ( this.applyHooks ) {
    await this.applyHooks(BEFORE_SAVE_HOOKS, showPreviewYamlContext);

....

async saveMachinePools(context?: string) {
   if (context === showPreviewYamlContext) {
      return;
   }   
```

I don't think we could make this a pattern everywhere due to the two  `await this.applyHooks(x, this.value);` requests in `impl.js`, but would be a succinct solution.

Happy also to go with the new changes in the PR (there's one minor comment)

Reply:
Thanks for the example, my mindset was originally stuck in trying to solve the problem solely in the `applyHooks` function instead of the applied function calls. 

I was a little apprehensive at first, but I can't argue with the succinctness of this approach. 

I think my issues stem from the "whatever" nature of `args` for these hooks; my initial approach didn't offer any enhancement in dealing with `args` at the end of the day 🙂